components/places/src/storage/history_metadata.rs

```diff
@@ -0,0 +1,992 @@
+        ORDER BY updated_at DESC
+        LIMIT {max_limit}",
+        common_select_sql = COMMON_METADATA_SELECT,
+        max_limit = MAX_QUERY_RESULTS
+    );
+    static ref GET_SINCE_SQL: String = format!(
+        "{common_select_sql}
+        WHERE updated_at >= :start
+        ORDER BY updated_at DESC
+        LIMIT {max_limit}",
+        common_select_sql = COMMON_METADATA_SELECT,
+        max_limit = MAX_QUERY_RESULTS
+    );
+    static ref QUERY_SQL: String = format!(
+        "{common_select_sql}
+        WHERE
+            url LIKE :query OR
+            title LIKE :query OR
+            search_term LIKE :query
+        ORDER BY total_view_time DESC
+        LIMIT :limit",
+        common_select_sql = COMMON_METADATA_SELECT
+    );
+}
+
+pub fn get_latest_for_url(db: &PlacesDb, url: &Url) -> Result<Option<HistoryMetadata>> {
+    let metadata = db.try_query_row(
+        GET_LATEST_SQL.as_str(),
+        &[(":url", &url.as_str())],
+        HistoryMetadata::from_row,
+        true,
+    )?;
+    Ok(metadata)
+}
+
+pub fn get_between(db: &PlacesDb, start: i64, end: i64) -> Result<HistoryMetadataList> {
+    let metadata = db.query_rows_and_then_named_cached(
+        GET_BETWEEN_SQL.as_str(),
+        rusqlite::named_params! {
+            ":start": start,
+            ":end": end,
+        },
+        HistoryMetadata::from_row,
+    )?;
+    Ok(HistoryMetadataList { metadata })
+}
+
+pub fn get_since(db: &PlacesDb, start: i64) -> Result<HistoryMetadataList> {
+    let metadata = db.query_rows_and_then_named_cached(
+        GET_SINCE_SQL.as_str(),
+        rusqlite::named_params! {
+            ":start": start
+        },
+        HistoryMetadata::from_row,
+    )?;
+    Ok(HistoryMetadataList { metadata })
+}
+
+pub fn query(db: &PlacesDb, query: &str, limit: i64) -> Result<HistoryMetadataList> {
+    let metadata = db.query_rows_and_then_named_cached(
+        QUERY_SQL.as_str(),
+        rusqlite::named_params! {
+            ":query": format!("%{}%", query),
+            ":limit": limit
+        },
+        HistoryMetadata::from_row,
+    )?;
+    Ok(HistoryMetadataList { metadata })
+}
+
+fn insert_metadata_in_tx(
+    tx: &PlacesTransaction<'_>,
+    place_id: Option<i64>,
+    search_query_entry: SearchQueryEntry,
+    parent_domain_entry: ParentDomainEntry,
+    metadata: HistoryMetadata,
+) -> Result<SyncGuid> {
+    // If HistoryMetadata already has a guid, that's an error.
+    assert!(metadata.guid.is_none());
+
+    // Heavy lifting around moz_places inserting (e.g. updating moz_origins, frecency, etc) is performed via triggers.
+    let places_id = match place_id {
+        Some(id) => id,
+        None => {
+            let sql = "INSERT INTO moz_places (guid, url, title, url_hash)
+               VALUES (:guid, :url, :title, hash(:url))";
+
+            let guid = SyncGuid::random();
+
+            // This normalizes urls, e.g. adding trailing slashes when they're missing.
```

Comment:
I think we should see if we can tweak the API rather than doing the sql here - I see 3 problems: (1) the query for an existing place ID is done outside the transaction, (2) we enforce a max title len (which we trim) and max URL len (which causes us to fail) that isn't done here and (3) IIUC the title for an existing moz_places record isn't updated when it doesn't match (and in that scenario you'd need to bump the change counter!)

You mentioned `new_page_info()` not taking a title param and the reason for this is that we l would need to do an update in the case where the page ID already exists. The "correct" API to use here would be something like [apply_observation](https://github.com/mozilla/application-services/blob/cd6c4e8a02f5339d87290aaa26709447b73c8ec8/components/places/src/storage/history.rs#L47) but that also adds a history record which we don't want - because presumably whatever is triggering this call to be made will also be adding a history visit too!

So I wonder if we can leverage that? If we can be certain that adding a history entry has already finished when this is called, we might be able to avoid even trying to create a new place and handling the title? IOW, can we just arrange to just bail of the place ID doesn't exist, because that should be impossible given there must also be an existing visit?

Otherwise I'll think about what shape the API should be (ie, how to split that above function in a way that splits `moz_places` from `moz_history_visits`) - probably just by adding a title param to `new_page_info()` and special-casing "might need to update title if the page already exists, but never if it doesn't" in `apply_observation()` - or maybe a new `ensure_page_info()` type function, which means you don't need to worry about any of that?

(Or maybe we just fix these problems here? I'll think about that :)

Reply:
For (1) and (2) I've pushed out the update. Both were oversights, I did think about doing them but promptly forgot :) Thanks for catching it.

For (3), we made a conscious choice to basically not worry about `title` updates. We're basically depending on the fact that there's a single source of these titles, and a certain sequencing in place, and an already existing title update mechanism in place. And in case we see a misalignment here, it's almost feels like a bug elsewhere (why didn't that title update persist already?) and basically doesn't matter for what we're building on top of this storage.

In practice, I doubt we'll actually end up inserting in moz_places/moz_origins (for parents) ourselves, in Fenix.

So I partially agree with "let's just not write to moz_places at all" logic (the paragraph above basically argues for that, between the lines).
But, that currently wouldn't work for iOS until they start using history APIs.

And, even in case of Fenix consuming this, not having fallback moz_places inserts starts to tie us into too many assumptions about the broader world.

So I think just ignoring title updates is a nice compromise. We've added comments in rust code and at the API boundary to make sure this behaviour is clear.